#### Logical Operator

In [1]:
1 and 0 

0

In [2]:
1 or 0

1

#### bitwise

In [ ]:
print(2^3)  #XOR
print(2&3)
print(~3) #NOT operator change 0 to 1 and 1 to 0

1
2


#### Assignment opeartor

#### Membership Operator
- in/ not in

In [5]:
print('D' in 'Delhi')


True


In [8]:
# Find the sun of a 3 digit number entered by user
value = int(input())
a = value%10
value = value//10
b = value%10
value = value//10
c = value%10
print(a+b+c)

15


In [10]:
def sum_num(num):
    total = 0
    while num>0:
        digit = num%10
        total+= digit
        num//=10
    return total
value = int(input())
print(sum_num(value))

15


In [11]:
# find the min of 3 given number
a = int(input("first num"))
b = int(input("Second num"))
c = int(input("Third num"))
print(min(a,b,c))

4


### module in python
- math
- keyword
- constant
- random


In [13]:
import datetime
print(datetime.datetime.now())

2025-08-18 19:41:22.420312


In [20]:
# program to print the table
num = int(input())
i = 1
while i<11:
    print(num*i)
    i +=1


11
22
33
44
55
66
77
88
99
110


In [ ]:
# sum of all digits to a given number 
# keep accepting the no from user till he enters a 0 and then find the average.

In [22]:
total = 0      # to keep sum of numbers
count = 0      # to count how many numbers entered

while True:
    num = int(input("Enter number (0 to stop): "))
    if num == 0:
        break              
    total += num
    count += 1

# avoid division by zero
if count > 0:
    avg = total / count
    print("Average:", avg)
else:
    print("No numbers entered!")


Average: 2.5


In [24]:
# Guessing game
import random


jackpot = random.randint(1,100)
guess = int(input())
count = 1
while guess!= jackpot:
    if guess< jackpot:
        print("Predict the no higher")
    else:
        print("Lower no")
    
    guess = int(input())
    count +=1
else:
    print('Correctly guessed')
    print(count)


Lower no
Predict the no higher
Predict the no higher
Lower no
Correctly guessed
5


In [23]:
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


#### The current population of a town is 10,000. The population of the town is increasing at a rate of 10% per year. How have you... You have to write a program to find out the population at the end of each of the last 10 years.

In [ ]:
population = 10000
for i in range(10,1,-1):
    print(i,population)
    population = population *1.1
# for last 10 years

10 10000
9 11000.0
8 12100.000000000002
7 13310.000000000004
6 14641.000000000005
5 16105.100000000008
4 17715.61000000001
3 19487.171000000013
2 21435.888100000015


In [31]:
for i in range(1,7):
    for j in range(1,7):
        if i ==j:
            print(i,j)
        

1 1
2 2
3 3
4 4
5 5
6 6


#### Pattern

In [38]:
rows = int(input('Enter no of rows: '))
for i in range(1,rows+1):
    for j in range(1,i+1):
        print('*', end = '')
    print()
        

*
**
***
****
*****


In [4]:
rows = int(input('Enter no of rows: '))
for i in range(1,rows+1):
    for j in range(1,i+1):
        print(j, end='')
    for k in range(i-1,0,-1):
        print(k,end='')
    print()

1
121
12321
1234321
123454321


In [8]:
# find prime no
start = int(input("Enter start of range: "))
end = int(input("Enter end of range: "))


for num in range(start, end + 1):
    if num > 1:   # primes are greater than 1
        is_prime = True
        for i in range(2, int(num**0.5) + 1):  # check divisibility till sqrt(num)
            if num % i == 0:
                is_prime = False
                break
        if is_prime:
            print(num, end=" ")


5 7 11 13 

In [9]:
start = int(input("Enter start of range: "))
end = int(input("Enter end of range: "))
for i in range(start,end):
    for j in range(2,i):
        if i%j==0:
            break
    else:
        print(i)

5
7
11
13


## String
- In python string is a collection of Unicode character

In [11]:
str = input("Enter a string")
str

'rahul'

In [14]:
for i in str:
    print(str)

rahul
rahul
rahul
rahul
rahul


In [15]:
for i in 'rahul':
    print(i)

r
a
h
u
l


- capitalize:- Only first char capitalize
- Title:- Capitalize first word
- Upper
- lower
- swapcase

#### format

In [17]:
age = 22
name = 'rahul'
print(f"the age is {age} and name is {name}")

the age is 22 and name is rahul


#### isalnum/ isalpha/ isdigit/ isidentifier

In [18]:
'rahul123'.isalnum()

True

## split/join

In [20]:
'hi my name is rahul'.split(' ')

['hi', 'my', 'name', 'is', 'rahul']

In [23]:
" ".join(['hi', 'my', 'name', 'is', 'rahul'])

'hi my name is rahul'

In [24]:
'hi my name is rahul'.replace('name', 'naam')

'hi my naam is rahul'

## List
- It is a datatype where you can store multiple items under 1 name. More technically, list sct like dynamic arrays which means you can add more items on the fly.

- list size is dynamic as compared to array (no define the size at first)
- list is heterogeneous (multiple type in a list)

In [30]:
# append
l = [1,2,3,4,5]
l.append(6)
l

[1, 2, 3, 4, 5, 6]

In [28]:
# extend
l = [1,2,3,4,5]
l.extend([6,7,8])
l

[1, 2, 3, 4, 5, 6, 7, 8]

In [33]:
# insert
l = [1,2,3,4,5]
l.insert(1,100)
l

[1, 100, 2, 3, 4, 5]

In [ ]:
# editing is possible with indexing and slicing
l = [1,2,3,4,5]
l[1] = 9
l

[1, 9, 3, 4, 5]

In [ ]:
# del
# remove
# pop
# clear (empty a list)
l = [1,2,3,4,5]
print(l)
del l
# del l[-1]
print(l)


[1, 2, 3, 4, 5]


NameError: name 'l' is not defined

In [ ]:
# count
l = [1,2,3,2,5,8,9,3,4,3]
l.count(3)

3

In [37]:
# reverse
l = [2,1,3,5,6]
l.reverse() # permanent change
l

[6, 5, 3, 1, 2]

In [38]:
l = [2,1,3,5,6]
print(l)
print(l.sort())
# sort make permanent change while shorted doesn't

[2, 1, 3, 5, 6]
None


#### List comprehension

In [40]:
lst = [i for i in range(1,11)]
lst

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

In [44]:
# print all the no divisible by 5 in range of 1 to 50
[i for i in range(1,51) if i%5==0]

[5, 10, 15, 20, 25, 30, 35, 40, 45, 50]

In [45]:
languages = ['java',' python',' c++', 'javascript']
[language for language in languages if language.startswith('j')]

['java', 'javascript']

## Tuples
- Same as list in python except it is immutable.

In [47]:
# empty tuple
t1 = ()
print(t1)

# single type tuple
t2 = ('hello',)
print(type(t2))

()
<class 'tuple'>


In [ ]:
# count (no of time it occur)
# index (index value)

## Sets
- Sets are unordered collection of items.
- Unordered 
- Mutable
- No duplicates
- Can't contain mutable data type

In [ ]:
# empty set
sets = {''} # s = set()
print(type(sets))

<class 'set'>


#### Accessing items
- Can't do indexing or slicing
- No changes can be done

#### Can add and del items in set

In [55]:
s = {1,2,4,5}
s.add(3)
s
# update
s.update([10])
s

{1, 2, 3, 4, 5, 10}

In [ ]:
del s

#### Set operations
- Union |
- Intersection &
- Difference -
- Symmetric difference ^ (exclude the same)

In [56]:
s1 = {1,2,3,4,5,6}
s2 = {7,8,9,10,11,12}
s1 | s2

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}

## Dictionary
- Store in key value pairs
- Mutable
- Indexing has no meaning
- No duplicate keys

#### Accessing items

## Functions

In [58]:
def is_even(num):
    if num%2==0:
        return 'Is even'
    else:
        return "Not even"
value = int(input())
print(is_even(value))

Is even


#### Type of argument
- While defining function it is called parameter and when giving value it became argument
- Default Argument
- Positional Argument
- Keyword Argument

*args, **kwargs

In [ ]:
def multiplying(*argum):
    # multiple no of variables no of non variable arguments can be taken with the help of *args
    product = 1
    for i in argum:
        product = product * i
    print(argum)
    return product

In [ ]:
# euclidean distance
import math
dis = math.sqrt((x2-x1)**2 + (y2-y1)**2)

In [ ]:
# Write a program to find the simple interest when the value of principle,rate of interest and time period is provided by the user.
SI = (Principle * R *T)/100

In [ ]:
# Write a program that will tell the number of dogs and chicken are there when the user will provide the value of total heads and legs.
def animals(heads, legs):
    # let dogs = d, chickens = c
    # equations:
    # d + c = heads
    # 4d + 2c = legs

    for dogs in range(heads + 1):
        chickens = heads - dogs
        if 4*dogs + 2*chickens == legs:
            return dogs, chickens
    
    return None, None  

heads = int(input("Enter number of heads: "))
legs = int(input("Enter number of legs: "))

dogs, chickens = animals(heads, legs)

if dogs is None:
    print("No valid solution exists!")
else:
    print("Dogs:", dogs)
    print("Chickens:", chickens)


Dogs: 2
Chickens: 2


In [9]:
# Write a program to find the sum of squares of first n natural numbers where n will be provided by the user.
def sum_sqr(num):
    total = 0
    for i in range(1,num+1):
        total += i**2
    return total
num = int(input("Enter the no"))
print(sum_sqr(num))

55


In [ ]:
# Given the first 2 terms of an Arithmetic Series.Find the Nth term of the series. Assume all inputs are provided by the user.
def nth_term_AP(a1, a2, n):
    d = a2 - a1         
    return a1 + (n - 1) * d

a1 = int(input("Enter the first term: "))
a2 = int(input("Enter the second term: "))
n = int(input("Enter the term number (n): "))

print(f"The {n}th term of the series is:", nth_term_AP(a1, a2, n))


The 7th term of the series is: 14


In [ ]:
# Given 2 fractions, find the sum of those 2 fractions.Take the numerator and denominator values of the fractions from the user.
num1 = int(input("Enter nu1: "))
den1 = int(input("Enter denominator1: "))
num2 = int(input("Enter nu2: "))
den2 = int(input("Enter denominator2: "))
frac1 = float(num1/den1)
frac2 = float(num2/den2)
final_sum = int(frac1+frac2)
print(final_sum)

"""
def add_fractions(num1, den1, num2, den2):
    # finding common denominator
    common_den = den1 * den2
    # adjusting numerators
    new_num1 = num1 * den2
    new_num2 = num2 * den1
    # sum of fractions
    final_num = new_num1 + new_num2
    return final_num, common_den

# input from user
num1 = int(input("Enter numerator1: "))
den1 = int(input("Enter denominator1: "))
num2 = int(input("Enter numerator2: "))
den2 = int(input("Enter denominator2: "))

num, den = add_fractions(num1, den1, num2, den2)
print(f"Sum of fractions = {num}/{den}")

"""

9


#### Given the height, width and breadth of a milk tank, you have to find out how many glasses of milk can be obtained? Assume all the inputs are provided by the user.



In [ ]:
def milk_glasses(height, width, breadth, glass_volume=250):
    # volume of milk tank (in cubic cm)
    tank_volume = height * width * breadth  
    
    # number of glasses
    glasses = tank_volume // glass_volume  
    
    return int(glasses)

# Taking user inputs
h = int(input("Enter height of milk tank (cm): "))
w = int(input("Enter width of milk tank (cm): "))
b = int(input("Enter breadth of milk tank (cm): "))

result = milk_glasses(h, w, b)
print("Total number of glasses of milk:", result)


Write a program that will give you in hand monthly salary after deduction on CTC - HRA(10%), DA(5%), PF(3%) and taxes deduction as below:

> Salary(Lakhs) : Tax(%)

*   Below 5 : 0%
*   5-10 : 10%
*   10-20 : 20%
*   aboove 20 : 30%

In [21]:
def in_hand(Salary):
    annual_salary = Salary*100000
    HRA = 0.10 * annual_salary
    DA = 0.05 * annual_salary
    PF = 0.03 * annual_salary
    salary_after_deduction = annual_salary - (HRA+DA+PF)
    if Salary < 5:
        tax_rate = 0
    elif Salary <= 10:
        tax_rate = 0.10
    elif Salary <= 20:
        tax_rate = 0.20
    else:
        tax_rate = 0.30

    tax = tax_rate * salary_after_deduction
    monthly_in_hand = (salary_after_deduction - tax)/12
    return monthly_in_hand

ctc = float(input("Enter your CTC in Lakhs: "))
print("Your Monthly In-Hand Salary is:", round(in_hand(ctc), 2))
    

Your Monthly In-Hand Salary is: 65600.0


In [ ]:
# Write a program that will take user input of cost price and selling price and determines whether its a loss or a profit.


In [23]:
# Display Fibonacci series up to 10 terms.
def fibonacci(n):
    a, b = 0, 1
    series = []
    for _ in range(n):
        series.append(a)
        a, b = b, a + b
    return series

print("Fibonacci Series:", fibonacci(10))


Fibonacci Series: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


In [ ]:
# Find the factorial of a given number.
def factorial(num):
    fact = 1
    if num == 0:
        return 1
    else:
        for i in range(1, num + 1):  
            fact *= i                
        return fact


#### Take a user input as integer N. Find out the sum from 1 to N. If any number if divisible by 5, then skip that number. And if the sum is greater than 300, don't need to calculate the sum further more. Print the final result. And don't use for loop to solve this problem.


In [ ]:
def sum_skip_5(N):
    total = 0
    i = 1
    while i <= N:
        if i % 5 == 0:   
            i += 1
            continue
        total += i
        if total > 300: 
            break
        i += 1
    return total

N = int(input("Enter a number: "))
print("Final Sum:", sum_skip_5(N))


In [25]:
#  Write a program that keeps on accepting a number from the user until the user enters Zero. Display the sum and average of all the numbers.
def accepting():
    total = 0
    count = 0
    
    while True:  
        N = int(input("Enter a number (0 to stop): "))
        
        if N == 0:
            break   
        
        total += N
        count += 1
    
    if count == 0:
        avg = 0
    else:
        avg = total / count
    
    print("Sum =", total)
    print("Average =", avg)

accepting()

        

Sum = 32
Average = 10.666666666666666


In [ ]:
# Find numbers divisible by 7 but not multiple of 5 between 2000 and 3200
result = []

for i in range(2000, 3201):  
    if i % 7 == 0 and i % 5 != 0:
        result.append(str(i))   # convert to string for joining later

print(",".join(result))


2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996,3003,3017,3024,3031,3038,3052,3059,3066,3073,3087,3094,3101,3108,3122,3129,3136,3143,3157,3164,3171,3178,3192,3199


In [27]:
# Write a program, which will find all such numbers between 1000 and 3000 (both included) such that each digit of the number is an even number. 
# The numbers obtained should be printed in a space-separated sequence on a single line.

result = []
for i in range(1000, 3001):
    all_even = True
    for digit in str(i):  # check each digit
        if int(digit) % 2 != 0:
            all_even = False
            break
    if all_even:
        result.append(str(i))

print(" ".join(result))


2000 2002 2004 2006 2008 2020 2022 2024 2026 2028 2040 2042 2044 2046 2048 2060 2062 2064 2066 2068 2080 2082 2084 2086 2088 2200 2202 2204 2206 2208 2220 2222 2224 2226 2228 2240 2242 2244 2246 2248 2260 2262 2264 2266 2268 2280 2282 2284 2286 2288 2400 2402 2404 2406 2408 2420 2422 2424 2426 2428 2440 2442 2444 2446 2448 2460 2462 2464 2466 2468 2480 2482 2484 2486 2488 2600 2602 2604 2606 2608 2620 2622 2624 2626 2628 2640 2642 2644 2646 2648 2660 2662 2664 2666 2668 2680 2682 2684 2686 2688 2800 2802 2804 2806 2808 2820 2822 2824 2826 2828 2840 2842 2844 2846 2848 2860 2862 2864 2866 2868 2880 2882 2884 2886 2888


###`Problem 11`: A robot moves in a plane starting from the original point (0,0). The robot can move toward UP, DOWN, LEFT and RIGHT with a given steps.
The trace of robot movement is shown as the following:
```
UP 5
DOWN 3
LEFT 3
RIGHT 2
!
```
> The numbers after the direction are steps.

> `!` means robot stop there.

**Please write a program to compute the distance from current position after a sequence of movement and original point.**

*If the distance is a float, then just print the nearest integer.*

Example:

`Input`:
```
UP 5
DOWN 3
LEFT 3
RIGHT 2
!
```
`Output`:
```
2
```

In [ ]:
import math

x, y = 0, 0  

while True:
    command = input().strip()
    if command == "!":   
        break

    direction, steps = command.split()
    # DOWn 5 is divided into direction = DOWN and steps = 5 
    steps = int(steps)

    if direction == "UP":
        y += steps
    elif direction == "DOWN":
        y -= steps
    elif direction == "LEFT":
        x -= steps
    elif direction == "RIGHT":
        x += steps

distance = math.sqrt(x**2 + y**2)
print(round(distance))   

In [31]:
# Write a program to print whether a given number is a prime number or not
def is_prime(N):
    for i in range(2, int(N**0.5)+ 1):
        if N %i==0:
            return False
    return True
n = int(input("Enter a number: "))

if is_prime(n):
    print(f"{n} is a prime number")
else:
    print(f"{n} is not a prime number")

5 is a prime number


#### `Problem 15`:Given two rectangles, find if the given two rectangles overlap or not. A rectangle is denoted by providing the x and y coordinates of two points: the left top corner and the right bottom corner of the rectangle. Two rectangles sharing a side are considered overlapping. (L1 and R1 are the extreme points of the first rectangle and L2 and R2 are the extreme points of the second rectangle).

Note: It may be assumed that the rectangles are parallel to the coordinate axis.

<img src='https://www.geeksforgeeks.org/wp-content/uploads/rectanglesOverlap.png' width='300' height='200'>

In [32]:
def do_overlap(L1, R1, L2, R2):
    # L1, R1, L2, R2 are tuples like (x, y)

    # If one rectangle is to the left of other
    if L1[0] > R2[0] or L2[0] > R1[0]:
        return False

    # If one rectangle is above the other
    if L1[1] < R2[1] or L2[1] < R1[1]:
        return False

    return True  # rectangles overlap

# --- Input example ---
L1 = (0, 10)
R1 = (10, 0)
L2 = (5, 5)
R2 = (15, -5)

if do_overlap(L1, R1, L2, R2):
    print("Rectangles Overlap")
else:
    print("Rectangles Do Not Overlap")


Rectangles Overlap


## OOP's

#### Class can have data and function only.

In [ ]:
# Syntax to create an object
# objectname = classname()
# L = list()

In [8]:
class Atm():
    # data are pin and balance
    # constructor(special function it gets automatically executed)
    def __init__(self):
        self.pin =  ''
        self.balance = 0
        print("Hola")
        self.menu() # so that menu gets automatically executed
        
    def menu(self):
        user_input = input("""
        Hi how can I help you?
        1. Press 1 to create pin
        2. Press 2 to change pin
        3. Press 3 to check balance
        4. Press 4 to Withdraw
        5. Press 5 to exit
              """)
        
        if user_input == '1':
            # create pin
            self.create_pin()
        
        elif user_input == '2':
            # change pin
            self.change_pin()
            
        elif user_input == '3':
            # check balance
            self.check_balance()
        
        elif user_input == '4':
            # withdraw
            self.withdraw()
        elif user_input == '5':
            print("Thank you for using the ATM. Have a nice day!")
            return
        else:
            print("Invalid input. Try again.")
            self.menu()
    
    def create_pin(self):
        user_pin = input('Enter your pin')
        self.pin = user_pin
        
        user_balance = int(input('Enter your balance'))
        self.balance = user_balance
        print("Pin created successfully")
        self.menu()
        
    def change_pin(self):
        old_pin = input('enter old pin')
        if old_pin == self.pin:
            # let hin change the pin
            new_pin = input('enter new pin')
            self.pin = new_pin
            print('pin change successfully')
            self.menu()
        else:
            print('pin galat hai re mamu')
            self.menu()
            
    def check_balance(self):
        user_pin = input('enter your pin')
        if user_pin == self.pin:
            print('your balance',self.balance)
            
        else:
            print('nikal pehli fursat se')
        self.menu()
            
    def withdraw(self):
        user_pin = input('enter your pin')
        if user_pin == self.pin:
            amount = int(input('enter the amount to withdraw'))
            if amount <= self.balance:
                self.balance = self.balance - amount
                print('withdraw successful, your balance is ', self.balance)
                
            else:
                print('not sufficient amount')
        else:
            print('pin ko wapus dekh bhai')
        self.menu()
    

In [10]:
# objectname = classname()
obj = Atm()

Hola
Thank you for using the ATM. Have a nice day!


In [4]:
print(type(obj))

<class '__main__.Atm'>


In [11]:
# Anything outside the class is called function and function inside the class is called method

#### Class diagram

#### Magic Method
- AKA dunder method


In [14]:
class Fraction():
    # parametrized 
    def __init__(self, x,y):
        self.num = x
        self.den = y
        
    def __str__(self):
        # use ths magic method to print
        return f"{self.num}/{self.den}"
        
        

In [15]:
frac1 = Fraction(3,4)
print(frac1)

3/4


## Encapsulation


In [16]:
# instance var - > python tutor
class Person:
    def __init__(self,name_input, country_input):
        self.name = name_input
        self.country = country_input
        
p1 = Person('nitish', 'india')
p2 = Person('steve', 'australia')

#### Class Relationships
- Aggregation (has a relationship)
- Inheritance

In [1]:
# example
class Customer:

  def __init__(self,name,gender,address):
    self.name = name
    self.gender = gender
    self.address = address # created a separate class called Address and pass the obj to the constructor

  def print_address(self):
    print(self.address._Address__city,self.address.pin,self.address.state)

  def edit_profile(self,new_name,new_city,new_pin,new_state):
    self.name = new_name
    self.address.edit_address(new_city,new_pin,new_state)

class Address:

  def __init__(self,city,pin,state):
      self.__city = city
      self.pin = pin
      self.state = state

  def get_city(self):
    return self.__city

  def edit_address(self,new_city,new_pin,new_state):
    self.__city = new_city
    self.pin = new_pin
    self.state = new_state

add1 = Address('gurgaon',122011,'haryana')
cust = Customer('nitish','male',add1)

cust.print_address()

cust.edit_profile('ankit','mumbai',111111,'maharastra')
cust.print_address()
# method example
# what about private attribute

gurgaon 122011 haryana
mumbai 111111 maharastra


#### Inheritance 

In [ ]:
#Inheritance
class User():
    def __init__(self):
        self.name = 'rahul'
    
    def login(self):
        print('login')
            
# Making child class -- Student from parent class -- User    
class Student(User):
    def enroll(self):
        print('enroll into the course')
# Method overriding if I try to make another constructor in the child class
u = User()
s = Student()

print(s.name)
s.login()
s.enroll()

rahul
login
enroll into the course


##### What gets inherited?

- Constructor
- Non Private Attributes
- Non Private Methods

In [9]:
# constructor example
class Phone():
    def __init__(self, price, brand , camera):
        print("Inside phone constructor")
        self.price = price
        self.brand = brand
        self.camera = camera
        
    def buy(Self):
        print("Buying a phone")
        
class SmartPhone(Phone):
    pass

s = SmartPhone(20000, 'Apple',13)
s.buy()
# Child don't have its constructor, so parent constructor is been called here

Inside phone constructor
Buying a phone


In [ ]:
# constructor example 2

class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

class SmartPhone(Phone):
    def __init__(self, os, ram):
        self.os = os
        self.ram = ram
        print ("Inside SmartPhone constructor")

s=SmartPhone("Android", 2)
s.os
s.brand
# Here child has its already existing constructor, so parent constructor were never created

Inside SmartPhone constructor


'Android'

In [12]:
# child can't access private members of the class

class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    #getter
    def show(self):
        print (self.__price)

class SmartPhone(Phone):
    def check(self):
        print(self.__price)

s=SmartPhone(20000, "Apple", 13)
s.show()
s.__price()

Inside phone constructor
20000


AttributeError: 'SmartPhone' object has no attribute '__price'

In [ ]:
# Method Overriding
# Parent and child has the same method name so child method will be executed
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class SmartPhone(Phone):
    def buy(self):
        print ("Buying a smartphone")

s=SmartPhone(20000, "Apple", 13)

s.buy()

### Super Keyword

In [13]:
# super is used to call the method of parent
class Phone():
    def __init__(self, price, brand, camera):
        print("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera
    def buy(self):
        print("Buying a phone")
        
class SmartPhone(Phone):
    def buy(self):
        print("Buying a smartphone")
        # Syntax to call parent ka buy method
        super().buy()
        
s = SmartPhone(20000, 'Apple', 13)
s.buy()


Inside phone constructor
Buying a smartphone
Buying a phone


In [ ]:
# super -> constructor
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

class SmartPhone(Phone):
    def __init__(self, price, brand, camera, os, ram):
        print('Inside smartphone constructor')
        super().__init__(price, brand, camera)
        self.os = os
        self.ram = ram
        print ("Inside smartphone constructor")

s=SmartPhone(20000, "Samsung", 12, "Android", 2)

print(s.os)
print(s.brand)

In [ ]:
# using super outside the class
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class SmartPhone(Phone):
    def buy(self):
        print ("Buying a smartphone")
        # syntax to call parent ka buy method
        super().buy()
        print(super().brand()) #cannot do this

s=SmartPhone(20000, "Apple", 13)
# super().buy() cannot be used here
s.buy()

- Super cannot access attributes
- Super cannot be called outside the class
- Super is used inside the child class

##### Inheritance in summary

- A class can inherit from another class.

- Inheritance improves code reuse

- Constructor, attributes, methods get inherited to the child class

- The parent has no access to the child class

- Private properties of parent are not accessible directly in child class

- Child class can override the attributes or methods. This is called method overriding

- super() is an inbuilt function which is used to invoke the parent class methods and constructor

In [14]:
class Parent:

    def __init__(self,num):
      self.__num=num

    def get_num(self):
      return self.__num

class Child(Parent):
  
    def __init__(self,num,val):
      super().__init__(num)
      self.__val=val

    def get_val(self):
      return self.__val
      
son=Child(100,200)
print(son.get_num())
print(son.get_val())

100
200


In [15]:
class Parent:
    def __init__(self):
        self.num=100

class Child(Parent):

    def __init__(self):
        super().__init__()
        self.var=200
        
    def show(self):
        print(self.num)
        print(self.var)

son=Child()
son.show()
# son and self is same, as self can access the non private member so num can also access that

100
200


In [16]:
class Parent:
    def __init__(self):
        self.__num=100

    def show(self):
        print("Parent:",self.__num)

class Child(Parent):
    def __init__(self):
        super().__init__()
        self.__var=10

    def show(self):
        print("Child:",self.__var)

obj=Child()
obj.show()

Child: 10


### Types of Inheritance

- Single Inheritance
- Multilevel Inheritance
- Hierarchical Inheritance
- Multiple Inheritance(Diamond Problem)
- Hybrid Inheritance

In [ ]:
# multilevel
class Product:
    def review(self):
        print ("Product customer review")

class Phone(Product):
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class SmartPhone(Phone):
    pass

s=SmartPhone(20000, "Apple", 12)

s.buy()
s.review()

In [17]:
# Hierarchical
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class SmartPhone(Phone):
    pass

class FeaturePhone(Phone):
    pass

SmartPhone(1000,"Apple","13px").buy()
FeaturePhone(10,"Lava","1px").buy()

Inside phone constructor
Buying a phone
Inside phone constructor
Buying a phone


In [ ]:
# Multiple
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class Product:
    def review(self):
        print ("Customer review")

class SmartPhone(Phone, Product):
    pass

s=SmartPhone(20000, "Apple", 12)

s.buy()
s.review()


In [18]:
# the diamond problem
# https://stackoverflow.com/questions/56361048/what-is-the-diamond-problem-in-python-and-why-its-not-appear-in-python2
class Phone:
    def __init__(self, price, brand, camera):
        print ("Inside phone constructor")
        self.__price = price
        self.brand = brand
        self.camera = camera

    def buy(self):
        print ("Buying a phone")

class Product:
    def buy(self):
        print ("Product buy method")

# Method resolution order
class SmartPhone(Phone,Product):
    pass
# here Phone is written first, so this will be executed at first

s=SmartPhone(20000, "Apple", 12)

s.buy()

Inside phone constructor
Buying a phone


In [19]:
class A:

    def m1(self):
        return 20

class B(A):

    def m1(self):
        return 30

    def m2(self):
        return 40

class C(B):
  
    def m2(self):
        return 20
obj1=A()
obj2=B()
obj3=C()
print(obj1.m1() + obj3.m1()+ obj3.m2())

70


In [20]:
class A:

    def m1(self):
        return 20

class B(A):

    def m1(self):
        val=super().m1()+30
        return val

class C(B):
  
    def m1(self):
        val=self.m1()+20
        # self.m1 is calling itself, infinite time so thats why it crashed
        return val
obj=C()
print(obj.m1())

RecursionError: maximum recursion depth exceeded

### Polymorphism
- When we have different method with same name, but with different features.

- Method Overriding:- Inheritance of child from parent
- Method Overloading:- Two method of same name in a class, but behaves differently
- Operator Overloading

In [ ]:
#Method overloading example
# Method overloading does not works in Python, But works in Java
# It helps to make it look more clean
class Shape():
    def area(self, radius):
        return 3.14*radius*radius
    
    def area(self, l,b):
        return l*b
    
s = Shape()
s.area(2)
s.area(3,4)

TypeError: Shape.area() missing 1 required positional argument: 'b'

In [ ]:
class Shape:
# Default argument can solve this problem
  def area(self,a,b=0):
    if b == 0:
      return 3.14*a*a
    else:
      return a*b

s = Shape()

print(s.area(2))
print(s.area(3,4))

In [22]:
#Operator overloading
print('hello' + 'world')
print(4 + 5)

helloworld
9


## Abstraction

In [ ]:
from abc import ABC,abstractmethod
class BankApp(ABC):
# BankApp is an abstract method -> which is an empty piece of code
  def database(self):
    print('connected to database')

  @abstractmethod
  def security(self):
    pass

  @abstractmethod
  def display(self):
    # it is a abstract method because no piece of code
    pass


In [29]:
class MobileApp(BankApp):

  def mobile_login(self):
    print('login into mobile')

  def security(self):
    print('mobile security')

  def display(self):
    print('display')

In [30]:
mob = MobileApp()

In [31]:
mob.security()

mobile security


In [ ]:
obj = BankApp()
#Can't create an object of the class
#  can only work with MobileApp

TypeError: Can't instantiate abstract class BankApp without an implementation for abstract methods 'display', 'security'

In [1]:
f = open('sample.txt','w')
f.write("Hello world")
f.close()

In [4]:
# Write multiline strings
f = open('sample.txt','w')
f.write("Hello world")
f.write('\nHow are U?')
f.close()

#### Namespaces

### Namespaces

A namespace is a space that holds names(identifiers).Programmatically speaking, namespaces are dictionary of identifiers(keys) and their objects(values)

There are 4 types of namespaces:
- Builtin Namespace
- Global Namespace
- Enclosing Namespace
- Local Namespace

In [5]:
# local and global -> same name
a = 2

def temp():
  # local var
  a = 3
  print(a)

temp()
print(a)

3
2


In [6]:
# local and global -> local does not have but global has
a = 2

def temp():
  # local var
  print(a)

temp()
print(a)


2
2


In [7]:
# local and global -> editing global
# Can't to changes directly
a = 2

def temp():
  # local var
  a += 1
  print(a)

temp()
print(a)

UnboundLocalError: cannot access local variable 'a' where it is not associated with a value

In [8]:
a = 2

def temp():
  # local var
  global a
  a += 1
  print(a)

temp()
print(a)

3
3


In [9]:
# local and global -> global created inside local
def temp():
  # local var
  global a
  a = 1
  print(a)

temp()
print(a)

1
1


In [10]:
# local and global -> function parameter is local
def temp(z):
  # local var
  print(z)

a = 5
temp(5)
print(a)
print(z)

5
5


NameError: name 'z' is not defined

In [11]:
# built in scope
print('hello')

hello


In [12]:
# How to see all the built-ins
import builtins
print(dir(builtins))

['ArithmeticError', 'AssertionError', 'AttributeError', 'BaseException', 'BaseExceptionGroup', 'BlockingIOError', 'BrokenPipeError', 'BufferError', 'BytesWarning', 'ChildProcessError', 'ConnectionAbortedError', 'ConnectionError', 'ConnectionRefusedError', 'ConnectionResetError', 'DeprecationWarning', 'EOFError', 'Ellipsis', 'EncodingWarning', 'EnvironmentError', 'Exception', 'ExceptionGroup', 'False', 'FileExistsError', 'FileNotFoundError', 'FloatingPointError', 'FutureWarning', 'GeneratorExit', 'IOError', 'ImportError', 'ImportWarning', 'IndentationError', 'IndexError', 'InterruptedError', 'IsADirectoryError', 'KeyError', 'KeyboardInterrupt', 'LookupError', 'MemoryError', 'ModuleNotFoundError', 'NameError', 'None', 'NotADirectoryError', 'NotImplemented', 'NotImplementedError', 'OSError', 'OverflowError', 'PendingDeprecationWarning', 'PermissionError', 'ProcessLookupError', 'RecursionError', 'ReferenceError', 'ResourceWarning', 'RuntimeError', 'RuntimeWarning', 'StopAsyncIteration', 'S

In [13]:
# renaming builtins
L = [1,2,3]
max(L)
def max():
    print('hello')
max(L)
# max is a already define function but due to LEGB rule it will check if max exist in global or not, if it  does then no need to check for the pre max function


TypeError: max() takes 0 positional arguments but 1 was given

### Decorators

A decorator in python is a function that receives another function as input and adds some functionality(decoration) to and it and returns it.

This can happen only because python functions are 1st class citizens.

There are 2 types of decorators available in python
- `Built in decorators` like `@staticmethod`, `@classmethod`, `@abstractmethod` and `@property` etc
- `User defined decorators` that we programmers can create according to our needs

In [ ]:
# Python are first class citizens
def func():
    return True
a = func()
a

In [ ]:
# not the decorator, calling decorator function with two arguments
def my_decorator(func, name):
    def wrapper():
        print("OOOOOOOOOOOOOOOOOOOOOO")
        func()
        name()
        print("LUNDLUNDLUND")
    return wrapper
def hello():
    print("Lawdapuri")
def display():
    print("Hello ji")
a = my_decorator(hello, display)
a()
b = my_decorator(display, hello) 
b()

OOOOOOOOOOOOOOOOOOOOOO
Lawdapuri
Hello ji
LUNDLUNDLUND
OOOOOOOOOOOOOOOOOOOOOO
Hello ji
Lawdapuri
LUNDLUNDLUND


In [16]:
def my_decorator(func):
    def wrapper():
        print("OOOOOOOOOOOOOOOOOOOOOO")
        func()
        print("LUNDLUNDLUND")
    return wrapper

@my_decorator
def hello():
    print("Lawdapuri")

@my_decorator
def display():
    print("Hello ji")

hello()
display()


OOOOOOOOOOOOOOOOOOOOOO
Lawdapuri
LUNDLUNDLUND
OOOOOOOOOOOOOOOOOOOOOO
Hello ji
LUNDLUNDLUND


In [ ]:
# from functools import wraps
# to display the docstring OR use this docstring inside the wrapper function
def decorator_function(original_function):
    def wrapper_function(*args, **kwargs):
        print("Wrapper executed before the original function.")
        return original_function(*args, **kwargs)
    return wrapper_function

@decorator_function
def display():
    """This function displays a message."""
    print("Display function executed.")

print(display.__name__)  # Output: wrapper_function
print(display.__doc__)   # Output: None



wrapper_function
This function displays a message.


In [17]:
# meaningful decorator
import time
def time_taken(func):
    def wrapper():
        start = time.time()
        func()
        print("Time taken is ", func.__name__, time.time()-start,'secs')
    return wrapper
@time_taken
def hello():
    print('hello world')
    time.sleep(2)
    
@time_taken
def display():
    print('Display')
    time.sleep(2)

hello()
display()
    
    
    

hello world
Time taken is  hello 2.001070976257324 secs
Display
Time taken is  display 2.000899076461792 secs


In [19]:
# Multiple arguments
import time
def time_taken(func):
    def wrapper(*args):
        start = time.time()
        func(*args)
        print("Time taken is ", func.__name__, time.time()-start,'secs')
    return wrapper
@time_taken
def hello():
    print('hello world')
    time.sleep(2)
    
@time_taken
def square(num):
    print(num**2)

hello()
square(2)
    
    
    

hello world
Time taken is  hello 2.00089693069458 secs
4
Time taken is  square 0.0 secs


In [27]:
def sanity_check(data_type): # data_type because decorator has its own type as int
    def outer_wrapper(func): # outer take data_type
        def inner_wrapper(*args): #inner take square
            if type(args[0]) == data_type:
                func(*args)
            else:
                print("EE na chaulbe")
        return inner_wrapper
    return outer_wrapper

@sanity_check(int)
def square(num):
    print(num**2)
@sanity_check(str)
def greet(name):
    print("Hello", name)
square(2)
greet('Rahul')


4
Hello Rahul


In [28]:
def outer_function(message):
    def inner_function():
        print(message)
    return inner_function

# Create a closure
my_closure = outer_function("Hello, Closure!")

# Call the closure - it still remembers 'message'
my_closure()

Hello, Closure!


#### What is a Iteration
- Iteration is a general term for taking each item of something, one after another. Any time you use a loop, explicit or implicit, to go over a group of items, that is iteration.

In [29]:
# Example
num = [1,2,3]

for i in num:
    print(i)

1
2
3


#### Iterator
- An Iterator is an object that allows the programmer to traverse through a sequence of data without having to store the entire data in the memory

In [ ]:
# Example
L = [x for x in range(1,10000)] # Iterable but not Iterator because it store in memory

#for i in L:
    #print(i*2)
    
import sys

print(sys.getsizeof(L)/1024)

x = range(1,10000000000)

#for i in x:
    #print(i*2)
    
print(sys.getsizeof(x)/1024)

83.1796875
0.046875


#### What is Iterable

- Iterable is an object, which one can iterate over

- It generates an Iterator when passed to iter() method.

In [32]:
# Example

L = [1,2,3]
type(L)


# L is an iterable
type(iter(L))

# iter(L) --> iterator

list_iterator

#### Point to remember
- Every Iterator is also and Iterable
- Not all Iterables are Iterators

#### Trick
- Every Iterable has an iter function
- Every Iterator has both iter function as well as a next function

In [33]:
a = 2
a

#for i in a:
    #print(i)
    
dir(a)

['__abs__',
 '__add__',
 '__and__',
 '__bool__',
 '__ceil__',
 '__class__',
 '__delattr__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__float__',
 '__floor__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getnewargs__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__index__',
 '__init__',
 '__init_subclass__',
 '__int__',
 '__invert__',
 '__le__',
 '__lshift__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rfloordiv__',
 '__rlshift__',
 '__rmod__',
 '__rmul__',
 '__ror__',
 '__round__',
 '__rpow__',
 '__rrshift__',
 '__rshift__',
 '__rsub__',
 '__rtruediv__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__truediv__',
 '__trunc__',
 '__xor__',
 'as_integer_ratio',
 'bit_count',
 'bit_length',
 'conjugate',
 'denominator',
 'from_bytes',
 'imag',
 'is_integer',
 

In [34]:
T = {1:2,3:4}
dir(T)

['__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__delitem__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__ior__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__reversed__',
 '__ror__',
 '__setattr__',
 '__setitem__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'clear',
 'copy',
 'fromkeys',
 'get',
 'items',
 'keys',
 'pop',
 'popitem',
 'setdefault',
 'update',
 'values']

#### For iterator

In [38]:
L = [1,2,3]
dir(L) # not iterator, because no __next__ available
# L is not an iterator


# iter() method is used to generate the iterable of a iterator 
iter_L = iter(L)
# iter_L is an iterator
# __iter__ and __next__ both available
dir(iter_L) 

['__class__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__length_hint__',
 '__lt__',
 '__ne__',
 '__new__',
 '__next__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__setstate__',
 '__sizeof__',
 '__str__',
 '__subclasshook__']

In [ ]:
# Create own range fun
# iterable
class mera_range:
    
    def __init__(self,start,end):
        self.start = start
        self.end = end
        
    def __iter__(self):
        return mera_iterator(self)
    

# iterator
class mera_iterator:
    
    def __init__(self,iterable_obj):
        self.iterable = iterable_obj
    
    def __iter__(self):
        return self
    
    def __next__(self):
        
        if self.iterable.start >= self.iterable.end:
            raise StopIteration
        
        current = self.iterable.start
        self.iterable.start+=1
        return current

## Generator
- Python generator are simple way to create iterators.
- Generator function don't have return, however have `yield`

#### The why?


In [39]:
L = [x for x in range(100000)]

#for i in L:
    #print(i**2)
    
import sys
sys.getsizeof(L)

x = range(10000000)

#for i in x:
    #print(i**2)
sys.getsizeof(x)

48

#### Simple example

In [42]:
def gen_demo():
    yield 'first statement'
    yield 'Second statement'
    yield 'Third statement'
    

In [45]:
gen = gen_demo()
print(next(gen))

first statement


### Practice Questions

##### `Problem 1` - Print the following pattern. Write a program to use for loop to print the following reverse number pattern.

```bash
5 4 3 2 1
4 3 2 1
3 2 1
2 1
1
```

In [62]:
for i in range(1,6):
    for j in range(6-i,0,-1):
        
        print(j, end=' ')
    print()

5 4 3 2 1 
4 3 2 1 
3 2 1 
2 1 
1 


##### `Problem 2`: Print the following pattern.

```bash
*
* *
* * *
* * * *
* * * * *
* * * *
* * *
* *
*
```

In [79]:
rows = int(input("Enter no of rows: "))
for i in range(1,rows+1):
    for j in range(0,i):
        print('*', end=' ')
    print()
for i in range(1,rows+1):
    for j in range(rows-i,0,-1):
        print('*', end=' ')
    print()

* 
* * 
* * * 
* * * * 
* * * * * 
* * * * 
* * * 
* * 
* 



##### `Problem 3`:Write a program to pring the following pattern

        *
      * * *
    * * * * *
   * * * * * * *
* * * * * * * * *


In [88]:
rows = 4
for i in range(1,rows+1):
    print(" "*rows, end=' ')
    print('* '*i)
    rows = rows-1
    

     * 
    * * 
   * * * 
  * * * * 


### `Problem 4`:Write a program to print the following pattern

1

2 1

3 2 1

4 3 2 1

5 4 3 2 1

In [95]:
n = 6
for i in range(1,6):
    for j in range(i,0,-1):
        print(j, end=' ')
    print()

1 
2 1 
3 2 1 
4 3 2 1 
5 4 3 2 1 


##### `Problem 5`: Write a Python Program to Find the Sum of the Series till the nth term:<br>
1 + x^2/2 + x^3/3 + … x^n/n<br>
n will be provided by the user

In [ ]:
x = int(input("Enter the value of x: "))
n = int(input("Enter the no of  term: "))
sum = 1
s = ''
print('1 + ', end ='')
for i in range(2,n+1):
    sum = sum + x**i/i
    s = s+ f"x^{i}/{i} + "
print(s[:-1])
print(sum)    

1 + x^2/2 +x^3/3 +x^4/4 +x^5/5 
22884.333333333332


##### `Problem 6`: The natural logarithm can be approximated by the following series.
If x is input through the keyboard, write a program to calculate the sum of the first seven terms of this series.


In [107]:
x = int(input("Enter the value of x: "))
n = 7
sum = 0
if x ==0:
    print('Output is infinite')
else:
    for i in range(1,8):
        sum = sum+ (1/i)*((x-i)/x)**i
print(s[:-1])
print(sum) 

x^2/2 +x^3/3 +x^4/4 +x^5/5 +(1/1)*((x-i)/x)^1 + (1/2)*((x-i)/x)^2 + (1/3)*((x-i)/x)^3 + (1/4)*((x-i)/x)^4 + (1/5)*((x-i)/x)^5 + (1/6)*((x-i)/x)^6 + (1/7)*((x-i)/x)^7 +
1.3736972428571432


##### `Problem 7` - Find the sum of the series upto n terms.
Write a program to calculate the sum of series up to n term. For example, if n =5 the series will become 2 + 22 + 222 + 2222 + 22222 = 24690. Take the user input and then calculate. And the output style should match which is given in the example.

**Example 1:**

`Input:`
```bash
5
```

`Output:`

```bash
2+22+222+2222+22222
Sum of above series is: 24690
```

In [ ]:
n = int(input("Enter the no of terms: "))
start = 2
sum = 0 
for i in range(0,n):
    if i < n-1:
        print(start, end='+')
    else:
         print(start)
    sum = sum+start
    start = start * 10 +2
print(sum)

2+22+222+2222+22222
24690


###### Problem 8: Write a program to print all the unique combinations of 1,2,3 and 4

`Output`:
```
1 2 3 4
1 2 4 3
1 3 2 4
1 3 4 2
1 4 2 3
1 4 3 2
2 1 3 4
2 1 4 3
2 3 1 4
2 3 4 1
2 4 1 3
.
.
and so on
```

In [108]:
for i in range(1,5):
    for k in range(1,5):
       for m in range(1,5):
            for l in range(1,5):
                print(i,j,m,l)
        

1 1 1 1
1 1 1 2
1 1 1 3
1 1 1 4
1 1 2 1
1 1 2 2
1 1 2 3
1 1 2 4
1 1 3 1
1 1 3 2
1 1 3 3
1 1 3 4
1 1 4 1
1 1 4 2
1 1 4 3
1 1 4 4
1 1 1 1
1 1 1 2
1 1 1 3
1 1 1 4
1 1 2 1
1 1 2 2
1 1 2 3
1 1 2 4
1 1 3 1
1 1 3 2
1 1 3 3
1 1 3 4
1 1 4 1
1 1 4 2
1 1 4 3
1 1 4 4
1 1 1 1
1 1 1 2
1 1 1 3
1 1 1 4
1 1 2 1
1 1 2 2
1 1 2 3
1 1 2 4
1 1 3 1
1 1 3 2
1 1 3 3
1 1 3 4
1 1 4 1
1 1 4 2
1 1 4 3
1 1 4 4
1 1 1 1
1 1 1 2
1 1 1 3
1 1 1 4
1 1 2 1
1 1 2 2
1 1 2 3
1 1 2 4
1 1 3 1
1 1 3 2
1 1 3 3
1 1 3 4
1 1 4 1
1 1 4 2
1 1 4 3
1 1 4 4
2 1 1 1
2 1 1 2
2 1 1 3
2 1 1 4
2 1 2 1
2 1 2 2
2 1 2 3
2 1 2 4
2 1 3 1
2 1 3 2
2 1 3 3
2 1 3 4
2 1 4 1
2 1 4 2
2 1 4 3
2 1 4 4
2 1 1 1
2 1 1 2
2 1 1 3
2 1 1 4
2 1 2 1
2 1 2 2
2 1 2 3
2 1 2 4
2 1 3 1
2 1 3 2
2 1 3 3
2 1 3 4
2 1 4 1
2 1 4 2
2 1 4 3
2 1 4 4
2 1 1 1
2 1 1 2
2 1 1 3
2 1 1 4
2 1 2 1
2 1 2 2
2 1 2 3
2 1 2 4
2 1 3 1
2 1 3 2
2 1 3 3
2 1 3 4
2 1 4 1
2 1 4 2
2 1 4 3
2 1 4 4
2 1 1 1
2 1 1 2
2 1 1 3
2 1 1 4
2 1 2 1
2 1 2 2
2 1 2 3
2 1 2 4
2 1 3 1
2 1 3 2
2 1 3 3
2 1 3 4
2 1 4 1


##### `Problem 9`: Write a program that will take a decimal number as input and prints out the binary equivalent of the number

In [109]:
n = int(input("Enter the no"))
binary=[]
while n>0:
    binary.append(n%2)
    n = n//2
    print(n,binary)
print(binary)

8 [1]
4 [1, 0]
2 [1, 0, 0]
1 [1, 0, 0, 0]
0 [1, 0, 0, 0, 1]
[1, 0, 0, 0, 1]


In [111]:
n = int(input())
binary = []
while n > 0:
    binary.append(n % 2)
    n = n // 2
print(binary[::-1])   


[1, 0, 0, 1, 0]


In [113]:
# Write a program that will take 2 numbers as input and prints the LCM and HCF of those 2 numbers
x = int(input("1st no"))
y = int(input('2nd no'))

if x>y:
    greater = x
else: 
    greater = y

while True:
    if (greater%x==0) and (greater%y==0):
        lcm = greater
        break
    greater = greater+1
print(lcm)
    

48


In [116]:
x = int(input("1st no"))
y = int(input('2nd no'))

if x<y:
    smaller = x
else:
    smaller = y
for i in range(1,smaller+1):
    if (x%i==0) and (y%i==0):
        hcf = i
print(i)
    

12


##### **Problem 11:** Create Short Form from initial character
Given a string create short form ofthe string from Initial character. Short form should be capitalised.

Example:

`Input:`
```
Data science mentorship program
```

`Output:`
```
DSMP

In [122]:
s = "Data science mentorship program".split()
res = ''.join(word[0].upper() for word in s)
print(res)



DSMP


In [123]:
inp = "Data science mentorship program"
res = ''
for i in inp.split():
    res = res + i[0].upper()
print(res)

DSMP


###`Problem 12`: Append second string in the middle of first string

`Input:`
```
campusx
data
```
`Output`:

```
camdatapusx
```

In [124]:
str1 = 'campusx'
str2 = 'data'
print(str1[0:int(len(str1)/2)] + str2 + str1[int(len(str1)/2):])

camdatapusx


##### ``Problem 13``:Given string contains a combination of the lower and upper case letters. Write a program to arrange the characters of a string so that all lowercase letters should come first.

Given:

`str1 = PyNaTive`

Expected Output:

`yaivePNT`

In [126]:
str1 = 'PyNaTive'
new_str1 = ''   # lowercase letters
new_str2 = ''   # uppercase letters

for i in str1:
    if i.islower():
        new_str1 += i
    else:
        new_str2 += i

print(new_str1 + new_str2)


yaivePNT


#### Take a alphanumeric string input and print the sum and average of the digits that appear in the string, ignoring all other characters.

Input:

`hel123O4every093`

Output:

```
Sum: 22
Avg: 2.75
```

In [ ]:
alpnum = input("Enter the alphanumeric character")
sum = 0
counter = 0
for i in alpnum:
    if i.isdigit():
        sum = sum+ int(i)
        counter = counter + 1
print(sum)
print(sum/counter)
print(counter)


22
3.142857142857143


###### `Problem 15:` Removal of all characters from a string except integers
`Given:`

```
str1 = 'I am 25 years and 10 months old'
```
`Expected Output:`
```
2510
```

In [135]:
str1 = 'I am 25 years and 10 months old'
res = ''
for i in str1:
    if i.isdigit():
        res = res+ i
print(res)

2510


##### `Problem 16`: Check whether the string is Symmetrical.

**Statement:** Given a string. the task is to check if the string is symmetrical or not. A string is said to be symmetrical if both the halves of the string are the same.

**Example 1:**

`Input`

```bash
khokho
```

`Output`

```bash
The entered string is symmetrical
```

In [144]:
s = input('Enter a string')
n = len(s)
if n % 2 == 0:
    print("Symmetric (equal halves)" if s[:n//2] == s[n//2:] else "Not symmetric (equal halves)")
else:
    # if you want to ignore the middle character for odd lengths:
    print("Symmetric (ignoring middle)"
          if s[:n//2] == s[n//2+1:] else "Not symmetric (equal halves)")

Symmetric (equal halves)


- `Problem 17`: Reverse words in a given String

**Statement:** We are given a string and we need to reverse words of a given string.

**Example 1:**

`Input:`

```bash
geeks quiz practice code
```

`Output:`

```bash
code practice quiz geeks
```

**Example 2:**

`Input:`

```bash
my name is laxmi
```

`Output:`

```bash
laxmi is name my
```

In [145]:
s = 'code practice quiz geeks'
l = []
for i in s.split():
    l.append(i)
l =  l[::-1]
print(" ".join(l))

geeks quiz practice code


In [147]:
str = 'mera naam mai bhulla'
str = str[::-1]
print(str)

alluhb iam maan arem


##### `Problem 18`: Find uncommon words from two Strings.

**Statement:** Given two sentences as strings **A** and **B**. The task is to return a list of all uncommon words. A word is uncommon if it appears exactly once in any one of the sentences, and does not appear in the other sentence. Note: A sentence is a string of space-separated words. Each word consists only of lowercase letters.

**Example 1:**

`Input:`

```bash
A = "apple banana mango"
B = "banana fruits mango"
```

`Output:`

```bash
['apple', 'fruits']
```

In [149]:
A = "apple banana mango"
B = "banana fruits mango"
l = []
for i in A.split():
    if i not in B and i not in l:
        l.append(i)
for j in B.split():
    if j not in A and j not in l:
        l.append(j)
print(l)

['apple', 'fruits']


##### `Problem 19`: Word location in String.

**Statement:** Find a location of a word in a given sentence.

**Example 1:**

`Input:`

```bash
Sentence: We can learn data science through campusx mentorship program.

word: campusx
```

`Output:`

```bash
Location of the word is 7.
```

Note- Don't use index/find functions

In [157]:
s = 'We can learn data science through campusx mentorship program.'
word = 'campusx'
pos = 0
for i in s.split():
    pos+=1
    if word in i:
        break
        
print(pos)

7


In [158]:
# Write a program that can remove all the duplicate characters from a string. User will provide the input.
s = 'abscdghafafcfhaf bacf abvcgad'
result = ''
for i in s:
    if i not in result:
        result = result + i
print(result)

abscdghf v
